File: app/src/main/java/ru/job4j/todolist/store/MemStore.java
package ru.job4j.todolist.store;

import android.content.Context;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import ru.job4j.todolist.Task;

public class MemStore implements IStore {
    private List<Task> tasks = new ArrayList<>();
    private static final MemStore INST = new MemStore();
    private int counter = 0;

    private MemStore() {
    }

    public MemStore(List<Task> tasks) {
        this.tasks = tasks;
    }

    public static MemStore getInstance(Context context) {
        return INST;
    }

    public List<Task> getTasks() {
        return this.tasks;
    }

    @Override
    public List<Task> getFilteredTasks(String filter) {
        return this.tasks.stream().filter(
                task -> task.getName().toLowerCase().contains(filter.toLowerCase()) ||
                        task.getDesc().toLowerCase().contains(filter.toLowerCase()) ||
                        task.getCreated().toLowerCase().contains(filter.toLowerCase()))
                .collect(Collectors.toList());
    }

    @Override
    public void addTask(String name,String description,String created, String photo) {
        this.tasks.add(new Task(counter++, name, description, created,null, photo));
    }

    @Override
    public void editTask(int id, String name,String description,String closed, String photo) {
        Task task = findTaskByID(id);
        task.setName(name);
        task.setDesc(description);
        task.setClosed(closed);
        task.setPhoto(photo);
    }

    @Override
    public void closeOrReopenTask(int id, String closed) {
        Task task = findTaskByID(id);
        task.setClosed(closed);
    }

    @Override
    public int deleteTask(int id) {
        int result = getPositionOfTaskById(id);
        this.tasks.remove(result);
        return result;
    }

    @Override
    public void deleteAllTasks() {
        this.tasks.clear();
    }

    @Override
    public Task findTaskByID(int id) {
        return this.tasks.get(getPositionOfTaskById(id));
    }

    @Override
    public int getPositionOfTaskById(int id) {
        for (int index = 0 ; index < tasks.size(); index ++) {
            if (tasks.get(index).getId() == id) {
                return index;
            }
        }
        return -1;
    }
}

File: app/src/main/java/ru/job4j/todolist/store/MemStore.kt
package ru.job4j.todolist.store

import android.content.Context
import ru.job4j.todolist.Task
import java.util.*
import java.util.stream.Collectors

class MemStore : IStore {
    private var tasks: MutableList<Task> = ArrayList()
    private var counter = 0

    companion object {
        private val INST = MemStore()
        fun getInstance(context: Context?): MemStore {
            return INST
        }
    }

    override fun getTasks(): List<Task> = tasks

    override fun getFilteredTasks(filter: String): List<Task> {
        return tasks.stream().filter { (_, name, desc, created) ->
            name.toLowerCase().contains(filter.toLowerCase()) ||
                    desc.toLowerCase().contains(filter.toLowerCase()) ||
                    created.toLowerCase().contains(filter.toLowerCase())
        }
                .collect(Collectors.toList())
    }

    override fun addTask(name: String, description: String, created: String, photo: String?) {
        tasks.add(Task(counter++, name, description, created, null, photo))
    }

    override fun editTask(id: Int, name: String, description: String, closed: String, photo: String) {
        val task = findTaskByID(id)
        task!!.name = name
        task.desc = description
        task.closed = closed
        task.photo = photo
    }

    override fun closeOrReopenTask(id: Int, closed: String?) {
        findTaskByID(id)?.closed = closed
    }

    override fun deleteTask(id: Int): Int =
            tasks.indexOf(findTaskByID(id)).apply {
                tasks.removeAt(this)
            }


    override fun deleteAllTasks() = tasks.clear()

    override fun findTaskByID(id: Int): Task? =
            tasks.firstOrNull { task -> task.id == id }
}